### ISSUE TITLE
NERACOOS' dataset inventory contains many outdated records

### ISSUE THREAD

**Issue by mwengren** [ioos/catalog#54]
At the NERACOOS DMAC review, we discovered that many datasets contained URLs that had been replaced by upstream metadata updates.  

Diagnose whether the issue is related to Solr index sync problems (ioos/catalog-ckan#158), other harvesting-related bug causing the CKAN database to be out of date, or a conflicting 'fileIdentifier' issue issue in source metadata that is preventing new records from harvesting.

- [ ] Notify Axiom when resolved to test NERACOOS station ingest process in Sensor Map

**Comment by ebridger** [comment]
@mwengren. I confirmed that the NERACOOS ISO WAF's do not contain unique 'fileIdentifier's. E.g. all the buoy B01 iso records for all 10 sensors or THREDDS NetCDF files have a fileIdentifier of edu.maine.B01.  This seems to date from Jun 23, 2017 which is when I switched our production TDS to https://hub.docker.com/r/unidata/thredds-docker/  iso records prior to that use the TDS urlPath as the file identifier.  I'm still investigating this.

**Comment by mwengren** [comment]
@ebridger, OK, keep us posted.  We are also looking into Catalog synchronization issues simultaneously, but resolving the duplicate fileidentifier issue will be necessary for any updates to make it through the Catalog.

**Comment by mwengren** [comment]
@ebridger, we have been troubleshooting the Catalog synchronization/harvesting issues over the past couple weeks.  We're not 100% through that, but we think things are in better shape than they were before.  

Because there were a lot of factors involved with the NERACOOS WAF update process, and potential update failures due to Catalog issues at the time, etc, I think it's possible that some of the NERACOOS service URLs may have updated properly from your new WAFs while others may not have.

The duplicate fileIndentifier issue will have to be resolved regardless within your Docker/THREDDS environment before we can harvest all of your records, but apart from that, it might make sense to do a full clear and reharvest of the NERACOOS WAFs to see if we get any different results with a healthier Catalog system.

Do you want to pursue that?  Are there legacy WAFs that should be removed from the Harvest Registry before we do a reharvest like this?  Or are they all hosting distinct data services (with to-be-differentiated fileIdentifiers)?

Currently, there are 178 NERACOOS datasets, with about 133 harvest failures, mostly due to the duplicate ID issue.

Just let us know if you want us to go ahead with this or how we can help resolve.

**Comment by kwilcox** [comment]
@ebridger Can you send the URL to your THREDDS server you are crawling to produce the ISO records? I had similar issues with the ISO translation for "parameter" based files (one file per parameter rather than one file per station) and I will attempt to remember what I did to fix this.

**Comment by ebridger** [comment]
@kwilcox http://www.neracoos.org/WAF/UMaine/Realtime/iso/
The non uniqueness is only across a particular buoy.

```
SOS_DSG_B01_AANDERAA_Realtime.xml
SOS_DSG_B01_ACCELEROMETER_Realtime.xml
SOS_DSG_B01_CTD1m_Realtime.xml
SOS_DSG_B01_CTD20m_Realtime.xml
SOS_DSG_B01_CTD50m_Realtime.xml
SOS_DSG_B01_MET_Realtime.xml
```

**Comment by kwilcox** [comment]
What is the URL to the THREDDS server that these ISO files were generated from? I can't seem to trace the ISO service URLS back to THREDDS URLs easily.

**Comment by ebridger** [comment]
Duh. Sorry.
 http://www.neracoos.org/thredds/UMO_SOS_realtime.html

**Comment by kwilcox** [comment]
The ISO files are using the combination of `naming_authority` and `id` to come up with the unique `fileIdentifier`. This was caused by updating THREDDS to a newer version that contained a new version of the ncISO translations.

ACDD states that:

> The combination of the "naming authority" and the "id" should be a globally unique identifier for the dataset.

You (and many others - myself included) split stations up into different datasets and each of the will require a unique `id` to be ACDD complaint. I would recommend you change the `id` attributes to something that identifies each specific dataset.

So for `A01`, instead of having `A01` as every single `id`, you could do:

* `A01-accel`
* `A01-met`
* `A01-ctd-1m`
* `A01-ctd-2m`
* `A01-oxygen-51m`

**Comment by kwilcox** [comment]
Here is an example from the CeNCOOS THREDDS Server. Each parameter measured at a station is split into its own dataset and given a unique `id`:

http://thredds.cencoos.org/thredds/catalog/cencoos/sensors/mlml_mlml_sea/data/catalog.html

`naming_authority: mlml`

`id`:
* `mlml-sea:sea_water_temperature`
* `mlml-sea:sea_water_practical_salinity`
* ...

**Comment by ebridger** [comment]
@kwilcox Thanks for clarifying this.  I had thought ncISO was using the THREDDS catalog ID, which also must be unique. Ran some tests and ncML can handle this w/o needing to update the NetCDF files. May just use the TDS ID as the id attribute via ncML.

**Comment by kwilcox** [comment]
You could do that but be aware that other tools will using `naming_authority` and `id` to create a URN/URI for the dataset and it will not be pretty if you use a THREDDS catalog ID.

**Comment by mwengren** [comment]
@ebridger If there's anything on the Catalog side we can do, let us know.  

Curious, would the THREDDS catalog ID you were mentioning for this dataset:

http://www.neracoos.org/thredds/UMO_SOS_realtime.html?dataset=SOS_DSG_A01_CTD1m_Realtime

be `SOS_DSG_A01_CTD1m_Realtime`?

So, since `naming authority = org.neracoos` the resulting ISO XML fileIdentifier would be:

`org.neracoos-SOS_DSG_A01_CTD1m_Realtime`?   

That seems reasonable.  Does that prevent ID collisions for your THREDDS Catalogs?

@kwilcox, trying to understand your last comment.  Do you mean that other tools connecting to their THREDDS server will have ugly IDs if their THREDDS catalog ID is more complex than something like `SOS_DSG_A01_CTD1m_Realtime`?  Just trying to understand for my own awareness the issues here, thanks.

**Comment by ebridger** [comment]
@mwengren, @kwilcox I think Kyle was saying that the id and naming authority for a NetCDF file belongs in the global attributes of the file which would be better than relying on the THREDDS catalog's ncML  to set them.  I think I agree and am working on such a system. But to confuse the matter further I came across this link: [conventions-for-observing-asset-identifiers/ioos-assets-v1-0.html ](https://ioos.github.io/conventions-for-observing-asset-identifiers/ioos-assets-v1-0.html) which adds a bunch of complexity and seems to imply long urn based id's.

**Comment by mwengren** [comment]
@ebridger Ok, sounds good.  Using the global attributes sounds logical to me.  I admit I am not entirely up on our Asset Identifier conventions.  I know @kwilcox had presented some recommended changes at the DMAC meeting last year, but I don't know if those were rolled into those guidelines or not.  

Also @abirger should be able to provide guidance on the status of the asset identifier conventions since he worked on that site.

**Comment by abirger** [comment]
@ebridger , 

NcSOS uses the following rules suggested in the ["Crosswalk from netCDF attributes to IOOS asset identifiers"](https://github.com/ioos/ioos-netcdf/blob/gh-pages/_docs/NetCDF-to-SOS%20Mappings_clean_2016-04-07a.xlsx):

 1. IOOS Asset Identifier is a URN that have the following structure: 

`urn:ioos:<asset_type>:<authority>:<label>:<component>:<discriminant><#functional_parameters>`,

where the field values are as follows:

|                       |     Value for platform       |       Value for sensor              | Mandatory/Optional
|:--------------- | :--------------------------- | :-----------------------| :---------------------------|
|asset_type             | Fixed “station”              | Fixed “sensor”                      |        M
|authority              | naming_authority             | naming_authority                    |        M
|label                  | platform_variable:short_name | platform_variable:short_name        |        M
component              |     N/A                      | geophysical_variable:standard_name  |        M
discriminant           |     N/A                      | geophysical_variable:discriminant   |        O
functional_parameters: |     N/A                      |                                     |
   a. cell_methods     |                              | geophysical_variable:cell_methods   |        O
   b. interval         |                              | geophysical_variable:interval       |        O
   c. vertical_datum*  |                              | geophysical_variable:vertical_datum |        O

\* the vertical_datum can also come from a variable with the standard_name `water_surface_reference_datum_altitude`, as defined in the CF spec
 2. It is required/strongly recommended that a description of each geophysical variable in netCDF file provides reference to the ancillary platform & instrument variables, e.g. 

`        sea_water_temperature:ancillary_variables = "platform_variable instrument_parameter_variable"` ;
`         sea_water_temperature:platform = "platform_variable1 platform_variable2"` ;
`         sea_water_temperature:instrument = "instrument1 instrument2 instrument3"` ;

3. For a platform, the following rules apply (in order of preference):
        a). it is required/strongly recommended that each ‘platform_variable#’ should have the `ioos_code` attribute (similar to existing attributes `wmo_code`, `nodc_code`, `imo_code`, etc.); the value of the `ioos_code` then serves as a complete IOOS Asset Identifier;
        b). alternatively, a value of `platform_variable#:short_name` attribute (e.g. `platform_variable_1:short_name = “carquinez”`) can be used for the `<label>` field of the IOOS Asset Identifier;
        c). if neither a) nor b) option is available, the value of a variable with a 'cf_role' attribute should be used for the `<label>` field;
        d). if neither of a) through c) options is available, the value of a global `id` attribute should be used for the `<label>` field.

Option `3d` seems to be exactly what you try to implement, with a combination of global attributes `naming_authority` and `id`. However, this option is the least preferable because it works well only if NetCDF file contains data collected from a single platform. Otherwise, the use of `id` for the `<label>` will end up in ambiguous SOS platform identifiers.

**Comment by mwengren** [comment]
@ebridger  This is an old issue that's lingering in our ioos/catalog repo.   Has the current version of the IOOS Catalog and any changes you made in TDS addressed the missing NERACOOS datasets/duplicate ID problem properly?  

We're having harvester issues in the CKAN part of the Catalog unfortunately, so can't see the NERACOOS CKAN harvest job [status](https://registry.ioos.us/jobs/show/NERACOOS?harvest=cVKJ2qFz6HN8AKGo0) the way you normally might, but if I log on to CKAN myself there's only about 4 records showing metadata errors in recent jobs.  Here's a summary of the error report, looks like mostly bad decimals in some datasets.  We can give you more info on these if it helps:

Count	Message
12	Element '{http://www.isotc211.org/2005/gco}Decimal': '9969209968386869000000000000000000000.0' is not a valid value of the atomic type 'xs:decimal'.
12	Element '{http://www.isotc211.org/2005/gco}Decimal': '' is not a valid value of the atomic type 'xs:decimal'.
2	Error parsing bounding box value: could not convert string to float:

Can we close this issue out?

**Comment by ebridger** [comment]
@mwengren Just seeing this now.   Just checked the UMaine WAF's and  all the files have unique fileIdentifier's so I think we are O.K. So close this. The bounding box is a separate issue which I'll look into.

**Comment by mwengren** [comment]
Ok, closing as we've resolved the NERACOOS identifier conflicts.
